cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs

```diff
@@ -0,0 +1,90 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Cardano.CLI.Types.MonadWarning
+--
+-- This module defines the 'MonadWarning' type class, which provides a common
+-- interface for monads that support reporting warning messages without
+-- aborting the computation (unlike with exceptions, Either, or MonadFail,
+-- which either fail or return a value).
+--
+-- It also includes two functions that instantiate it into either a 'MonadIO'
+-- ('monadWarningInMonadIO') or a 'StateT' monad with a '[String]' as state
+-- ('monadWarningInStateT') respectively.
+--
+-- In the case of 'MonadIO', warnings are printed to 'stderr'.
+-- In the case of 'StateT', with a '[String]' state, warnings are added to the
+-- list in the state.
+--
+-- By using the 'MonadWarning' type class, users can write code that remains
+-- agnostic to the specific monad in which it operates, and to easily change
+-- it at a later stage if necessary.
+--
+-- Example usage:
+--
+-- @
+-- computeWithWarning :: (MonadWarning m) => Int -> m Int
+-- computeWithWarning x = do
+--   when (x < 0) $ reportIssue "Input value is negative!"
+--   return (x * 2)
+--
+-- -- Using 'IO' monad to perform computation and report warnings.
+-- main :: IO ()
+-- main = do
+--     result <- monadWarningInMonadIO $ computeWithWarning (-4)
+--     putStrLn $ "Result: " ++ show result
+-- @
+-----------------------------------------------------------------------------
+
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs #-}
+
+module Cardano.CLI.Types.MonadWarning
+  ( MonadWarning(..)
+  , eitherToWarning
+  , monadWarningInMonadIO
+  , monadWarningInStateT
+  ) where
+
+import           Cardano.Prelude (MonadIO (liftIO), MonadState (state), Print (hPutStrLn), StateT,
+                   stderr)
+
+-- | Type class for monads that support reporting warnings.
+class Monad m => MonadWarning m where
+    -- | Report a warning issue.
+    reportIssue :: String -- ^ The warning message to report.
+                -> m ()   -- ^ The action that reports the warning.
+
+-- | Wrapper newtype for 'MonadIO' with 'MonadWarning' instance.
+-- We need to have wrapper to avoid overlapping instances.
+newtype WarningMonadIO m a = WarningMonadIO (m a)
+    deriving (Functor, Applicative, Monad, MonadIO)
+
+-- | This instance prints the issue to the 'stderr'.
+instance MonadIO m => MonadWarning (WarningMonadIO m) where
+    reportIssue :: String -> WarningMonadIO m ()
+    reportIssue issue = liftIO (hPutStrLn stderr issue)
+
+-- | Convert a 'MonadWarning' into a 'MonadIO' by reporting
+-- warnings to 'stderr'.
+monadWarningInMonadIO :: WarningMonadIO m a -> m a
```

Comment:
The standard in mtl seems to call such functions using a verb `runXXX` e.g. `runExcept`. I think just
```haskell
newtype WarningIO m a = WarningIO { runWarningIO :: m a }
    deriving (Functor, Applicative, Monad, MonadIO)

newtype WarningStateT m a = WarningStateT { runWarningStateT :: StateT [String] m a }
    deriving (Functor, Applicative, Monad, MonadState [String])
```
without this and `monadWarningInStateT` functions would be enough. Also, the name `Monad` in the newtypes seems redundant.